### PR TITLE
Add SGSparseMatrix equals and operator==

### LINKED ISSUE #4071: implement SGSparseMatrix::equals
See 80808c65d886309d64d2b1d944cd84e3ae822451
We want the same for our sparse matrices

Easy entrance task, that allows you to get exposed to some internals of shogun. It is mostly copy paste, but requires some understanding of the above patch.

### PR DESCRIPTION
**Pull request by vinx13**
This fixes #4071

### CODE REVIEW

File: src/shogun/lib/SGSparseMatrix.h [hunk 1]
```diff
@@ -187,6 +187,13 @@ template <class T> class SGSparseMatrix : public SGReferencedData
 		/** sort the indices of the sparse matrix such that they are in ascending order */
 		void sort_features();
 
+		/** Pointer identify comparison.
+		 *  @return true iff number of vectors and features and pointer are equal
+		 */
+		inline bool operator==(const SGSparseMatrix<T>& other) const;
```
Comment (lisitsyn) [inline src/shogun/lib/SGSparseMatrix.h]: `inline` does not change anything as implementation is in .cpp

**Comment by lisitsyn** [comment]
Thanks for the patch!

Please take a look at travis job that is failing https://travis-ci.org/shogun-toolbox/shogun/jobs/328007339

The patch needs some codestyle fixes.

**Comment by vinx13** [comment]
I tried to fix some code style but I got into trouble with git branch. I'll try to fix it then.

File: src/shogun/lib/SGSparseMatrix.cpp [hunk 2]
```diff
@@ -282,6 +282,40 @@ template<class T> void SGSparseMatrix<T>::from_dense(SGMatrix<T> full)
...
+{
+	// same instance
+	if (*this == other)
+		return true;
+
+	// different size
+	if (num_vectors != other.num_vectors || num_features != other.num_features)
+		return false;
+
+	for (int i = 0; i < num_vectors; i++)
```
Comment (karlnapf) [inline src/shogun/lib/SGSparseMatrix.cpp]: auto i : range(num_vectors)

File: src/shogun/lib/SGSparseMatrix.h [hunk 3]
```diff
@@ -187,7 +187,15 @@ template <class T> class SGSparseMatrix : public SGReferencedData
...
 		void sort_features();
 
-protected:
+		/** Pointer identify comparison.
+		 *  @return true iff number of vectors and features and pointer are
+		 * equal
+		 */
+		bool operator==(const SGSparseMatrix<T>& other) const;
+
+		bool equals(const SGSparseMatrix<T>& other) const;
```
Comment (karlnapf) [inline src/shogun/lib/SGSparseMatrix.h]: needs doxygen docs

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 4]
```diff
@@ -286,3 +286,47 @@ TEST(SGSparseMatrix, transposed_square_matrix)
...
+
+TEST(SGSparseMatrix, equals_same_shape)
+{
+	const index_t number_of_features = 2;
+	const index_t number_of_vectors = 2;
+
+	SGSparseMatrix<float64_t> m1(number_of_vectors, number_of_features);
+	SGSparseMatrix<float64_t> m2(number_of_vectors, number_of_features);
+
+	for (int i = 0; i < number_of_vectors; i++)
```
Comment (karlnapf) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: auto...range...

**Comment by karlnapf** [comment]
The style thing can easily be fixed via looking at the travis output, it gives you a command that formats all files corretly: https://travis-ci.org/shogun-toolbox/shogun/jobs/328007339#L720

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 5]
```diff
@@ -286,3 +287,47 @@ TEST(SGSparseMatrix, transposed_square_matrix)
...
+	EXPECT_FALSE(m2.equals(m1));
+}
+
+TEST(SGSparseMatrix, equals_different_shape)
+{
+	SGSparseMatrix<float64_t> m1(2, 2);
+	SGSparseMatrix<float64_t> m2(2, 3);
+	EXPECT_FALSE(m1.equals(m2));
+	EXPECT_FALSE(m2.equals(m1));
+}
```
Comment (karlnapf) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: we also need a test for the pointer comparison. Sorry I didnt spot that before

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 6]
```diff
@@ -331,3 +331,18 @@ TEST(SGSparseMatrix, equals_different_shape)
 	EXPECT_FALSE(m1.equals(m2));
 	EXPECT_FALSE(m2.equals(m1));
 }
+
+TEST(SGSparseMatrix, pointe_equal)
```
Comment (karlnapf) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: typo

File: tests/unit/lib/SGSparseMatrix_unittest.cc [hunk 7]
```diff
@@ -331,3 +331,18 @@ TEST(SGSparseMatrix, equals_different_shape)
...
+	SGSparseMatrix<float64_t> m2(2, 3);
+	EXPECT_FALSE(m1 == m2);
+	EXPECT_TRUE(m1 == m1);
+	EXPECT_TRUE(m2 == m2);
+
+	SGSparseMatrix<float64_t> m3(2, 2);
+	EXPECT_FALSE(m1 == m3);
+
+	SGSparseMatrix<float64_t> m4(m1);
+	EXPECT_TRUE(m1 == m4);
```
Comment (karlnapf) [inline tests/unit/lib/SGSparseMatrix_unittest.cc]: Although I really like that you though of this case, I think this is a test for the copy constructor and not for ==.
Would remove it

Comment (vinx13) [reply tests/unit/lib/SGSparseMatrix_unittest.cc]: Removed. Please take a look.

File: src/shogun/lib/SGSparseMatrix.h [hunk 8]
```diff
@@ -187,8 +187,20 @@ template <class T> class SGSparseMatrix : public SGReferencedData
 		/** sort the indices of the sparse matrix such that they are in ascending order */
 		void sort_features();
 
-protected:
+		/** Pointer identify comparison.
```
Comment (karlnapf) [inline src/shogun/lib/SGSparseMatrix.h]: typo: identity

**Comment by karlnapf** [comment]
Good stuff. Thanks!

**Pull request merged**